src/main/java/io/microprofile/config/ConfigProvider.java

```diff
@@ -13,28 +13,163 @@
+ * That way it is possible to overwrite configuration with lower importance from outside.</p>
+ *
+ * <p>It is also possible to register custom {@link ConfigSource}s to
+ * flexibly extend the configuration mechanism. An example would be to pick up configuration values
+ * from a database table./p>
+ *
+ * <p>Example usage:
+ *
+ * <pre>
+ *     String restUrl = ConfigProvider.getConfig().getValue("myproject.some.remote.service.url");
+ *     Integer port = ConfigProvider.getConfig().getValue("myproject.some.remote.service.port", Integer.class);
+ * </pre>
+ *
+ * </p>
  *
+ * @author <a href="mailto:struberg@apache.org">Mark Struberg</a>
+ * @author <a href="mailto:rmannibucau@apache.org">Romain Manni-Bucau</a>
+ * @author <a href="mailto:emijiang@uk.ibm.com">Emily Jiang</a>
  */
-public interface ConfigProvider {
-
-	/**
-	 * Create a builder so that custom config resources can be added.
-	 * @return an instance {@link ConfigBuilder}
-	 */
-	ConfigBuilder builder();
-	/**
-	 * Get the {@link Config} after aggregating on multiple {@link ConfigSource}s
-	 * @return the {@link Config} containing all config properties
-	 */
-	Config getConfig();
+public class ConfigProvider {
+
+    private static final ConfigProviderResolver instance = ConfigProviderResolver.instance();
+
+    /**
+     * Provide a {@link Config} based on all {@link ConfigSource}s
+     * of the current Thread Context ClassLoader (TCCL)
+     *
+     * <p>There is exactly a single Config instance per ClassLoader</p>
+     */
+    public static Config getDefaultConfig() {
+        return instance.getDefaultConfig();
+    }
+
+    /**
+     * Provide a {@link Config} based on all {@link ConfigSource}s
+     * of the specified Thread Context ClassLoader (TCCL)
+     *
+     * <p>There is exactly a single Config instance per ClassLoader</p>
+     */
+    public static Config getConfig(ClassLoader cl) {
+        return instance.getConfig(cl);
+    }
+
+
+    /**
+     * Create a fresh {@link ConfigBuilder} instance.
+     * This ConfigBuilder will initially contain no
+     * {@link ConfigSource} nor any {@link Converters}.
+     * Those have to be added manually.
+     *
+     * The ConfigProvider will not manage the Config instance internally
+     */
+    public static ConfigBuilder emptyBuilder() {
+        return instance.emptyBuilder();
+    }
+    /**
+     * Create a {@link ConfigBuilder} instance.
+     * This builder contains the default {@link ConfigSource} and the {@link Converters}.
+     * @return the ConfigBuilder instance
+     */
+    public static ConfigBuilder builder() {    	
+    	return instance.builder();
+    }
+
+    /**
+     * A {@link Config} normally gets released if the ClassLoader it represents gets destroyed.
+     * Invoke this method if you like to destroy the Config prematurely.
+     */
+    public static void releaseConfig(Config config) {
+        instance.releaseConfig(config);
```

Comment:
Oh that is also 1:1 taken from my proposal it seems. Shouldn't we slowly go back and start over there? Anyway, the reason is that:
There are 2 ways to bind a Config to an application:

1.) very 'loose'. See https://github.com/struberg/javaConfig/blob/master/impl/src/main/java/org/apache/geronimo/config/DefaultConfigProvider.java#L33
By having a *very* weak binding we can be sure to not ever have a memory leak (at least not caused by config). 
What most people do is to just use a WeakHashMap. But for anything ClassLoader related that doesn't add any benefit. Because the key 'locks' the value from getting freed. So it's totally worth nothing. By having the value itself also a WeakReference it really gets garbage collected if the Config is not used anymore.
And here is also the problem: If you don't have any reference to a Config object, well then your config will get gced the next time mem runs full... And on the next usage it needs to get built up again. 
That _might_ be a bit too eager :)

2.) strongly bound. Instead of doing the WeakReference tricks we can also just use a Map of Application (or CL) -> Config. Because in that case we have to 'manually' free the Config anyway if the application shuts down or gets undeployed. Therefore you need a releaseConfig method...

3.) Another 'trick'. It's a bit dirty and I don't really like it myself. But I simply had no better idea...
Apache DeltaSpikes config is *not* immutable! And we use that to register ConfigSources via CDI Events (ProcessAnnotatedType, @Config, etc). Remember that the ServiceLoader stuff gets triggered when the first config value gets requested. This is usually *very* early in the bootstrapping. At least long before any CDI config could get scanned. At this early stage we simply don't know the full List of ConfigSources. A possible solution would be to have an empty ConfigSourceProvider (https://github.com/struberg/javaConfig/blob/master/api/src/main/java/io/microprofile/config/spi/ConfigSourceProvider.java). Initially this would return an empty List (should be Iterable finally). Then during class scanning all the info gets collected. 
In a certain phase (e.g. AfterDeploymentValidation) one could invoke releaseConfig() to reset all the Config. And the next usage of getConfig would then pick up the freshly collected ConfigSources. 

Again: I do not really like that trick - it feels like an utter hack. Happy to hear about your ideas how that might get solved. 

LieGrue,
strub